extensions/msal-node-extensions/src/dpapi-addon/dpapi_win.cpp

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+ // Implementation referenced from https://github.com/bradhugh/node-dpapi
+
+#include <node.h>
+#include <nan.h>
+#include <Windows.h>
+#include <dpapi.h>
+#include <functional>
+#include <iostream>
+#include <string>
+
+v8::Local<v8::String> CreateUtf8String(v8::Isolate* isolate, char* strData)
+{
+	return v8::String::NewFromUtf8(isolate, strData, v8::NewStringType::kNormal).ToLocalChecked();
+}
+
+void ProtectDataCommon(bool protect, Nan::NAN_METHOD_ARGS_TYPE info)
+{
+	v8::Isolate* isolate = info.GetIsolate();
+
+	if (info.Length() != 3) {
+		isolate->ThrowException(v8::Exception::RangeError(
+			CreateUtf8String(isolate, "3 arguments are required")));
+	}
+
+	if (info[0]->IsNullOrUndefined() || !info[0]->IsUint8Array())
+	{
+		isolate->ThrowException(v8::Exception::TypeError(
+			CreateUtf8String(isolate, "First argument, data, must be a valid Uint8Array")));
+	}
+
+	if (!info[1]->IsNull() && !info[1]->IsUint8Array())
+	{
+		isolate->ThrowException(v8::Exception::TypeError(
+			CreateUtf8String(isolate, "Second argument, optionalEntropy, must be null or an ArrayBuffer")));
+	}
+
+	if (info[2]->IsNullOrUndefined() || !info[2]->IsString())
+	{
+		isolate->ThrowException(v8::Exception::TypeError(
+			CreateUtf8String(isolate, "Third argument, scope, must be a string")));
+	}
+
+	DWORD flags = 0;
+	if (!info[2]->IsNullOrUndefined())
+	{
+		v8::String::Utf8Value strData(isolate, info[2]);
+		std::string scope(*strData);
+		if (stricmp(scope.c_str(), "LocalMachine") == 0)
+		{
+			flags = CRYPTPROTECT_LOCAL_MACHINE;
+		}
+	}
+
+	auto buffer = node::Buffer::Data(info[0]);
+	auto len = node::Buffer::Length(info[0]);
+
+	DATA_BLOB entropyBlob;
+	entropyBlob.pbData = nullptr;
+	if (!info[1]->IsNull())
+	{
+		entropyBlob.pbData = reinterpret_cast<BYTE*>(node::Buffer::Data(info[1]));
+		entropyBlob.cbData = node::Buffer::Length(info[1]);
```

Comment:
Do you have a guarantee that both of these buffers are null-terminated?
Or, does `node::Buffer::Length` have some other form of safety check to prevent an overrun?

Reply:
The data here (info[1]) is the user data that we are passing in to the windows API `CryptProtectData()`. My theory is that node's `ArrayBufferView()` implementation - used in the internal implementation of `node::Buffer:Data()`, should be adding the null terminator but I asked @sangonzal to run some experiments confirming that. If not, we may need to add the null termination ourselves.